Pipeline requests near the chain tip too, with care

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1650 


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 0a90f237a76d5895e0623e9a3f1a57b724512b13
  :round_pushpin: **Pipeline requests near the chain tip too, with care**
    This is a bit tricky because, as soon as we enter the `tip - k` area,
  any 'RequestNext' may yield a 'RollBackward'. The strategy is however
  quite simple and boils down to two cases:

  - Either we have to rollback within blocks we have just collected but
    not yet sent for processing.

  - Either, we have to rollback beyond this.

  In the former case, we can simply discard blocks from what we have
  collected and continue as normal. In the latter, we have to forward
  the rollback requests to the wallet engine, but we have to keep on
  collecting the results for the requests we have already pipelined!

  We do so by stashing the results in a queue, so that they'd be
  available for the next 'RequestNext' command from the wallet.

  This effectively extend pipelining up to the point where we have
  caught up with the tip and start collecting requests one by one, in a
  calmly manner. This is visible from the logs:

  ```
  ...
  Applying blocks [192.13732 ... 192.14731]
  Applying blocks [192.14732 ... 192.15731]
  Applying blocks [192.15732 ... 192.15977]
  Applying blocks [192.15978 ... 192.15978]
  Applying blocks [192.15979 ... 192.15979]
  Applying blocks [192.15980 ... 192.15980]
  ```


# Comments

<!-- Additional comments or screenshots to attach if any -->

This should fix #1650 

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking commented on lib/byron/1-percent.timelog:
```diff
@@ -0,0 +1,4320 @@
+1589384183.268922344s -0
+1589384183.34260801s 1000
+1589384183.401558201s 2000
+1589384183.462808633s 3000
```
Shouldn't be committed!

piotr-iohk replied:
That's for boosting GH statistics ;P

![Screenshot from 2020-05-14 16-53-04](https://user-images.githubusercontent.com/42900201/81949742-75952b80-9603-11ea-98e0-d2c53d98c3d2.png)

KtorZ replied:
I might have abused `git add lib` :smile: ...

KtorZ committed:
Pipeline requests near the chain tip too, with care

  This is a bit tricky because, as soon as we enter the `tip - k` area,
  any 'RequestNext' may yield a 'RollBackward'. The strategy is however
  quite simple and boils down to two cases:

  - Either we have to rollback within blocks we have just collected but
    not yet sent for processing.

  - Either, we have to rollback beyond this.

  In the former case, we can simply discard blocks from what we have
  collected and continue as normal. In the latter, we have to forward
  the rollback requests to the wallet engine, but we have to keep on
  collecting the results for the requests we have already pipelined!

  We do so by stashing the results in a queue, so that they'd be
  available for the next 'RequestNext' command from the wallet.

  This effectively extend pipelining up to the point where we have
  caught up with the tip and start collecting requests one by one, in a
  calmly manner. This is visible from the logs:

  ```
  ...
  Applying blocks [192.13732 ... 192.14731]
  Applying blocks [192.14732 ... 192.15731]
  Applying blocks [192.15732 ... 192.15977]
  Applying blocks [192.15978 ... 192.15978]
  Applying blocks [192.15979 ... 192.15979]
  Applying blocks [192.15980 ... 192.15980]
  ```

Commit:
Regenerate nix

Anviking commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -340,24 +355,61 @@ chainSyncWithBlocks (Quantity epochStability) fromTip queue =
+        -- receiving future requests, we'll simply read them from the stash!
+        , P.recvMsgRollBackward = \point _tip ->
+            case rollback point blocks of
+                [] -> do -- b)
```
It probably doesn't matter, but I think you can't differentiate between:
- Rolling back to the latest persisted block (`[]`, but no `respond (Rollback point)` needed)
- Rolling back to one before the latest persisted block (rollback needed)

KtorZ replied:
Yes, this would induce a rollback to where the wallet already is at, which is a no-op for the wallet and database. So, in principle, there's nothing wrong from it. Yet indeed, the client is blind to what the latest persisted block is so it can't really make this distinction.

Anviking reviewed (approved):
lgtm.
I guess testing that everything works if the chain forks in the middle of collecting responses, etc, would be nice, but not sure how one would do that.

rvl commented on lib/byron/src/Cardano/Wallet/Byron/Network.hs:
```diff
@@ -326,9 +326,10 @@ mkWalletClient
-    nodeToClientProtocols NodeToClientProtocols
+    -> m (NetworkClient m)
+mkWalletClient bp chainSyncQ = do
+    stash <- atomically newTQueue
```
This could maybe be called "responseBuffer"

KtorZ replied:
done in: 0446953c8f6a888a6af5d4d19895d8b918a5884f :+1:

rvl reviewed (approved):
Looks good to me.

Would the design be simpler if we changed network layer from pull (nextBlocks) to push (onNextBlock)?

It would also be good to allocate some time to unit testing the cardano-node network layers because they are not totally simple any more.

KtorZ committed:
rename 'stash' to 'responseBuffer' for clarity
